src/System.Private.ServiceModel/src/System/IdentityModel/Tokens/KeyInfoSerializer.cs

```diff
@@ -0,0 +1,370 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IdentityModel.Selectors;
+using System.Runtime;
+using System.ServiceModel.Security;
+using System.Xml;
+using TokenEntry = System.ServiceModel.Security.WSSecurityTokenSerializer.TokenEntry;
+using DiagnosticUtility = System.ServiceModel.DiagnosticUtility;
+
+namespace System.IdentityModel.Tokens
+{
+    /// <summary>
+    /// Abstract class for SecurityKeyIdentifierClause Serializer.
+    /// </summary>
+    internal class KeyInfoSerializer : SecurityTokenSerializer
+    {
+        private readonly List<SecurityTokenSerializer.KeyIdentifierEntry> _keyIdentifierEntries;
+        private readonly List<SecurityTokenSerializer.KeyIdentifierClauseEntry> _keyIdentifierClauseEntries;
+        private readonly List<SecurityTokenSerializer.SerializerEntries> _serializerEntries;
+        private readonly List<TokenEntry> _tokenEntries;
+
+        private DictionaryManager _dictionaryManager;
+        private bool _emitBspRequiredAttributes;
+        private SecurityTokenSerializer _innerSecurityTokenSerializer;
+
+        /// <summary>
+        /// Creates an instance of <see cref="SecurityKeyIdentifierClauseSerializer"/>
+        /// </summary>
+        public KeyInfoSerializer(bool emitBspRequiredAttributes)
+            : this(emitBspRequiredAttributes, new DictionaryManager(), XD.TrustDec2005Dictionary, null)
+        {
+        }
+
+        public KeyInfoSerializer(
+            bool emitBspRequiredAttributes,
+            DictionaryManager dictionaryManager,
+            TrustDictionary trustDictionary,
+            SecurityTokenSerializer innerSecurityTokenSerializer ) :
+            this( emitBspRequiredAttributes, dictionaryManager, trustDictionary, innerSecurityTokenSerializer, null )
+        {
+        }
+
+        public KeyInfoSerializer(
+            bool emitBspRequiredAttributes,
+            DictionaryManager dictionaryManager,
+            TrustDictionary trustDictionary,
+            SecurityTokenSerializer innerSecurityTokenSerializer,
+            Func<KeyInfoSerializer, IEnumerable<SerializerEntries>> additionalEntries)
+        {
+            _dictionaryManager = dictionaryManager;
+            _emitBspRequiredAttributes = emitBspRequiredAttributes;
+            _innerSecurityTokenSerializer = innerSecurityTokenSerializer;
+
+            _serializerEntries = new List<SecurityTokenSerializer.SerializerEntries>();
+
+            // Issue #31 in progress
+            //serializerEntries.Add(new XmlDsigSep2000(this));
+            //serializerEntries.Add(new XmlEncApr2001(this));
+            //serializerEntries.Add(new WSTrust(this, trustDictionary));
+
```

Comment:
Would it be possible to add skeleton classes that throw for these? That way it's less likely this will get missed as missing these items to serialize could be really difficult bug to track down.